File: gps_tracker/ios/Runner/NativeGpsBuffer.swift
import Foundation
import CoreLocation

/// Stores GPS points captured natively (outside Flutter) as JSONL (one JSON object per line).
/// Flutter reads and drains this buffer on resume via MethodChannel.
///
/// Thread safety: all operations are serialized on a private DispatchQueue.
/// Migration: on first drain(), any leftover data in the old UserDefaults key is merged.
class NativeGpsBuffer {
    static let shared = NativeGpsBuffer()

    private let fileName = "native_gps_buffer.jsonl"
    private let queue = DispatchQueue(label: "ca.trilogis.gpstracker.native-gps-buffer")

    // Old UserDefaults key (for one-time migration)
    private let oldKey = "native_gps_buffer_points"

    private var fileURL: URL {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return dir.appendingPathComponent(fileName)
    }

    private func ensureDirectory() {
        let dir = fileURL.deletingLastPathComponent()
        if !FileManager.default.fileExists(atPath: dir.path) {
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        }
    }

    func save(location: CLLocation, shiftId: String) {
        queue.sync {
            let point: [String: Any] = [
                "shift_id": shiftId,
                "latitude": location.coordinate.latitude,
                "longitude": location.coordinate.longitude,
                "accuracy": location.horizontalAccuracy,
                "altitude": location.altitude,
                "speed": max(0, location.speed),
                "heading": max(0, location.course),
                "captured_at": Int(location.timestamp.timeIntervalSince1970 * 1000),
                "source": "native_slc"
            ]

            guard let data = try? JSONSerialization.data(withJSONObject: point),
                  var line = String(data: data, encoding: .utf8) else { return }

            line += "\n"

            ensureDirectory()

            let path = fileURL.path
            if !FileManager.default.fileExists(atPath: path) {
                FileManager.default.createFile(atPath: path, contents: nil)
            }

            guard let handle = FileHandle(forWritingAtPath: path) else { return }
            defer { handle.closeFile() }
            handle.seekToEndOfFile()
            if let lineData = line.data(using: .utf8) {
                handle.write(lineData)
            }
        }
    }

    func drain() -> String {
        return queue.sync {
            var points: [[String: Any]] = []

            // --- One-time migration from old UserDefaults ---
            let oldJson = UserDefaults.standard.string(forKey: oldKey) ?? "[]"
            if oldJson != "[]",
               let oldData = oldJson.data(using: .utf8),
               let oldArray = try? JSONSerialization.jsonObject(with: oldData) as? [[String: Any]] {
                points.append(contentsOf: oldArray)
                UserDefaults.standard.set("[]", forKey: oldKey)
            }

            // --- Read JSONL file ---
            let path = fileURL.path
            if FileManager.default.fileExists(atPath: path) {
                if let content = try? String(contentsOfFile: path, encoding: .utf8) {
                    let lines = content.components(separatedBy: "\n")
                    for line in lines {
                        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
                        if trimmed.isEmpty { continue }
                        if let data = trimmed.data(using: .utf8),
                           let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                            points.append(obj)
                        }
                        // else: corrupted/truncated line — skip
                    }
                }
                try? FileManager.default.removeItem(atPath: path)
            }

            // Return JSON array string (same format as before)
            if let data = try? JSONSerialization.data(withJSONObject: points),
               let json = String(data: data, encoding: .utf8) {
                return json
            }
            return "[]"
        }
    }

    func count() -> Int {
        return queue.sync {
            let path = fileURL.path
            guard FileManager.default.fileExists(atPath: path),
                  let content = try? String(contentsOfFile: path, encoding: .utf8) else {
                return 0
            }
            return content.components(separatedBy: "\n")
                .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
                .count
        }
    }
}

File: gps_tracker/android/app/src/main/kotlin/ca/trilogis/gpstracker/NativeGpsBuffer.kt
package ca.trilogis.gpstracker

import android.content.Context
import android.location.Location
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedWriter
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStreamWriter
import java.util.UUID

/**
 * Stores GPS points captured natively (outside Flutter) as JSONL (one JSON object per line).
 * Flutter reads and drains this buffer on resume via MethodChannel.
 *
 * Migration: On first drain() after app update, any leftover data in the old
 * SharedPreferences store is merged into the result and the old key is cleared.
 */
object NativeGpsBuffer {
    private const val FILE_NAME = "native_gps_buffer.jsonl"

    // Old SharedPreferences keys (for one-time migration)
    private const val OLD_PREFS_NAME = "native_gps_buffer"
    private const val OLD_KEY_POINTS = "points"

    fun save(context: Context, shiftId: String, location: Location) {
        val point = JSONObject().apply {
            put("shift_id", shiftId)
            put("latitude", location.latitude)
            put("longitude", location.longitude)
            put("accuracy", location.accuracy.toDouble())
            put("altitude", location.altitude)
            put("speed", location.speed.toDouble())
            put("heading", location.bearing.toDouble())
            put("captured_at", System.currentTimeMillis())
            put("source", "native_rescue")
            val clientId = UUID.nameUUIDFromBytes(
                "$shiftId:${location.time}".toByteArray()
            ).toString()
            put("client_id", clientId)
        }

        try {
            val file = File(context.filesDir, FILE_NAME)
            FileOutputStream(file, true).use { fos ->
                OutputStreamWriter(fos, Charsets.UTF_8).use { writer ->
                    BufferedWriter(writer).use { bw ->
                        bw.write(point.toString())
                        bw.newLine()
                        bw.flush()
                    }
                }
            }
        } catch (_: Exception) {
            // IOException / disk full — skip this point, next alarm will retry
        }
    }

    fun drain(context: Context): String {
        val result = JSONArray()

        // --- One-time migration from old SharedPreferences ---
        try {
            val prefs = context.getSharedPreferences(OLD_PREFS_NAME, Context.MODE_PRIVATE)
            val oldJson = prefs.getString(OLD_KEY_POINTS, "[]") ?: "[]"
            if (oldJson != "[]") {
                val oldArray = JSONArray(oldJson)
                for (i in 0 until oldArray.length()) {
                    result.put(oldArray.getJSONObject(i))
                }
                prefs.edit().putString(OLD_KEY_POINTS, "[]").apply()
            }
        } catch (_: Exception) {
            // Migration failed — old data may be lost, but don't crash
        }

        // --- Read JSONL file ---
        val file = File(context.filesDir, FILE_NAME)
        if (file.exists()) {
            try {
                file.bufferedReader(Charsets.UTF_8).useLines { lines ->
                    for (line in lines) {
                        val trimmed = line.trim()
                        if (trimmed.isEmpty()) continue
                        try {
                            result.put(JSONObject(trimmed))
                        } catch (_: Exception) {
                            // Corrupted/truncated line (e.g. SIGKILL mid-write) — skip
                        }
                    }
                }
            } catch (_: Exception) {
                // File read error — return whatever we have from migration
            }
            file.delete()
        }

        return result.toString()
    }

    fun count(context: Context): Int {
        val file = File(context.filesDir, FILE_NAME)
        if (!file.exists()) return 0
        return try {
            file.bufferedReader(Charsets.UTF_8).useLines { lines ->
                lines.count { it.trim().isNotEmpty() }
            }
        } catch (_: Exception) {
            0
        }
    }
}
